component/src/main/java/com/celements/struct/DefaultStructDataService.java

```diff
@@ -0,0 +1,109 @@
+import org.slf4j.LoggerFactory;
+import org.xwiki.component.annotation.Component;
+import org.xwiki.component.annotation.Requirement;
+import org.xwiki.component.phase.Initializable;
+import org.xwiki.component.phase.InitializationException;
+import org.xwiki.velocity.VelocityManager;
+import org.xwiki.velocity.XWikiVelocityException;
+
+import com.celements.common.reflect.ReflectiveInstanceSupplier;
+import com.celements.convert.bean.BeanClassDefConverter;
+import com.celements.convert.bean.XObjectBeanConverter;
+import com.celements.model.access.IModelAccessFacade;
+import com.celements.model.classes.ClassDefinition;
+import com.celements.model.context.ModelContext;
+import com.celements.model.object.xwiki.XWikiObjectFetcher;
+import com.celements.model.util.ModelUtils;
+import com.celements.rights.access.exceptions.NoAccessRightsException;
+import com.celements.struct.classes.TableClass;
+import com.celements.struct.classes.TableColumnClass;
+import com.celements.struct.table.ColumnConfig;
+import com.celements.struct.table.TableConfig;
+import com.google.common.base.Optional;
+import com.google.common.base.Strings;
+import com.xpn.xwiki.doc.XWikiDocument;
+import com.xpn.xwiki.objects.BaseObject;
+
+@Component
+public class DefaultStructDataService implements StructDataService, Initializable {
+
+  private static Logger LOGGER = LoggerFactory.getLogger(DefaultStructDataService.class);
+
+  @Requirement
+  private ModelUtils modelUtils;
+
+  @Requirement
+  private IModelAccessFacade modelAccess;
+
+  @Requirement
+  private ModelContext context;
+
+  @Requirement
+  private VelocityManager velocityManager;
+
+  @Requirement(XObjectBeanConverter.NAME)
+  private BeanClassDefConverter<BaseObject, TableConfig> tableConverter;
+
+  @Requirement(XObjectBeanConverter.NAME)
+  private BeanClassDefConverter<BaseObject, ColumnConfig> columnConverter;
+
+  @Requirement(TableClass.CLASS_DEF_HINT)
+  private ClassDefinition tableClass;
+
+  @Requirement(TableColumnClass.CLASS_DEF_HINT)
+  private ClassDefinition columnClass;
+
+  @Override
+  public void initialize() throws InitializationException {
+    tableConverter.initialize(tableClass);
+    tableConverter.initialize(new ReflectiveInstanceSupplier<>(TableConfig.class));
+    columnConverter.initialize(columnClass);
+    columnConverter.initialize(new ReflectiveInstanceSupplier<>(ColumnConfig.class));
+  }
+
+  @Override
+  public String evaluateVelocityText(XWikiDocument doc, String text) throws NoAccessRightsException,
+      XWikiVelocityException {
+    checkNotNull(doc);
+    VelocityContext vContext = (VelocityContext) velocityManager.getVelocityContext().clone();
+    vContext.put("doc", modelAccess.getApiDocument(doc));
+    return evaluateVelocityText(doc, text, vContext);
+  }
+
+  @Override
+  public String evaluateVelocityText(String text) throws XWikiVelocityException {
+    return evaluateVelocityText(context.getDoc(), text, velocityManager.getVelocityContext());
```

Comment:
because the context is not manipulated here, unlike line 78

Reply:
Without cloning the Velocity-Script can change context-variables and it has an impact on outside scripts (global behaviour). With cloning variable-changes in the Velocity-scripts behave like local variable. I think this is irritating and must be handled transparently in the same way for the Velocity-developers. Please decide if variables are always behaving like globale variables or like local variables and synchronize the implementations.

Reply:
I see your point. I decided on local variable behaviour and added the clone since I don't see a use case to manipulate outside scripts from within StructEdit-Macro-Fields.